core/src/main/java/org/apache/iceberg/deletes/RoaringPositionBitmap.java

```diff
@@ -0,0 +1,309 @@
+      buffer.putInt(index);
+      bitmaps[index].serialize(buffer);
+    }
+  }
+
+  private void allocateBitmapsIfNeeded(int requiredLength) {
+    if (bitmaps.length < requiredLength) {
+      if (bitmaps.length == 0 && requiredLength == 1) {
+        this.bitmaps = new RoaringBitmap[] {new RoaringBitmap()};
+      } else {
+        RoaringBitmap[] newBitmaps = new RoaringBitmap[requiredLength];
+        System.arraycopy(bitmaps, 0, newBitmaps, 0, bitmaps.length);
+        for (int index = bitmaps.length; index < requiredLength; index++) {
+          newBitmaps[index] = new RoaringBitmap();
+        }
+        this.bitmaps = newBitmaps;
+      }
+    }
+  }
+
+  /**
+   * Deserializes a bitmap from a buffer, assuming the portable serialization format.
+   *
+   * @param buffer the buffer to read from
+   * @return a new bitmap instance with the deserialized data
+   */
+  public static RoaringPositionBitmap deserialize(ByteBuffer buffer) {
+    validateByteOrder(buffer);
+
+    // the bitmap array may be sparse with more elements than the number of read bitmaps
+    int remainingBitmapCount = readBitmapCount(buffer);
+    List<RoaringBitmap> bitmaps = Lists.newArrayListWithExpectedSize(remainingBitmapCount);
+    int lastKey = 0;
+
+    while (remainingBitmapCount > 0) {
+      int key = readKey(buffer, lastKey);
+
+      // fill gaps as the bitmap array may be sparse
+      while (lastKey < key) {
+        bitmaps.add(new RoaringBitmap());
+        lastKey++;
+      }
+
+      RoaringBitmap bitmap = readBitmap(buffer);
+      bitmaps.add(bitmap);
+
+      lastKey++;
+      remainingBitmapCount--;
+    }
+
+    return new RoaringPositionBitmap(bitmaps.toArray(EMPTY_BITMAP_ARRAY));
+  }
+
+  private static void validateByteOrder(ByteBuffer buffer) {
+    Preconditions.checkArgument(
+        buffer.order() == ByteOrder.LITTLE_ENDIAN,
+        "Roaring bitmap serialization requires little-endian byte order");
+  }
+
+  private static int readBitmapCount(ByteBuffer buffer) {
+    long bitmapCount = buffer.getLong();
+    Preconditions.checkArgument(
+        bitmapCount >= 0 && bitmapCount <= Integer.MAX_VALUE,
+        "Invalid bitmap count: %s",
+        bitmapCount);
+    return (int) bitmapCount;
+  }
+
+  private static int readKey(ByteBuffer buffer, int lastKey) {
+    int key = buffer.getInt();
+    Preconditions.checkArgument(key >= 0, "Invalid unsigned key: %s", key);
+    Preconditions.checkArgument(key >= lastKey, "Keys must be sorted in ascending order");
+    return key;
+  }
+
+  private static RoaringBitmap readBitmap(ByteBuffer buffer) {
+    try {
+      RoaringBitmap bitmap = new RoaringBitmap();
+      bitmap.deserialize(buffer);
+      buffer.position(buffer.position() + bitmap.serializedSizeInBytes());
+      return bitmap;
+    } catch (IOException e) {
+      throw new UncheckedIOException(e);
+    }
+  }
+
+  // extracts high 32 bits from a 64-bit position
+  private static int highBytes(long pos) {
```

Comment:
keyBytes?

Reply:
I tend to agree here, since you need to read the javadoc to know that these are the key bytes, so might be worth renaming the method to indicate that

Reply:
Or just `key`?

Reply:
The problem is that input in many methods is called `position` and represents 64-bit positions. If we want to rename this method to  `key` (which I agree makes sense), we will have to consider renaming `lowBytes` too.

Let me think.